Writing tools animations on iOS don't have the second part of the animation.

megangardner opened the pull request:
#### da92dcfbdd3a6ec697b543983f82e4ea370cd851
<pre>
Writing tools animations on iOS don&apos;t have the second part of the animation.
<a href="https://bugs.webkit.org/show_bug.cgi?id=279161">https://bugs.webkit.org/show_bug.cgi?id=279161</a>
<a href="https://rdar.apple.com/132367375">rdar://132367375</a>

Reviewed by Aditya Keerthi.

iOS animations have a completely different shape than mac
animations. Also, we weren&apos;t correctly determining if something
was the final text replacement and bailing out of that animations.
We also needed to ensure that smart replies still worked correctly.

* Source/WebCore/dom/Document.h:
(WebCore::Document::setPlaceholderTextIndicatorData):
(WebCore::Document::placeholderTextIndicatorData):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::insertTextPlaceholder):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::addSourceTextAnimationForActiveWritingToolsSession):
(WebCore::ChromeClient::addDestinationTextAnimationForActiveWritingToolsSession):
* Source/WebCore/page/TextAnimationTypes.h:
* Source/WebCore/page/writing-tools/WritingToolsController.h:
* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::compositionSessionDidFinishReplacement):
(WebCore::WritingToolsController::compositionSessionDidReceiveTextWithReplacementRangeAsync):
(WebCore::WritingToolsController::compositionSessionDidReceiveTextWithReplacementRange):
* Source/WebKit/Shared/TextAnimationTypes.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _addTextAnimationForAnimationID:withData:]):
(-[WKWebView _enableSourceTextAnimationAfterElementWithID:]):
(-[WKWebView _enableFinalTextAnimationForElementWithID:]):
(toWKTextAnimationType): Deleted.
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::addTextAnimationForAnimationIDWithCompletionHandler):
(WebKit::WebPageProxy::storeDestinationCompletionHandlerForAnimationID):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setUpInteraction]):
(-[WKContentView updateUnderlyingTextVisibilityForTextAnimationID:visible:completionHandler:]):
(-[WKContentView callCompletionHandlerForAnimationID:]):
(-[WKContentView callCompletionHandlerForAnimationID:completionHandler:]):
(toWKTextAnimationType):
(-[WKContentView addTextAnimationForAnimationID:withData:]):
(-[WKContentView addTextAnimationForAnimationID:withStyleType:]): Deleted.
* Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
(-[WKTextAnimationManager restoreTextAnimationType]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::addTextAnimationForAnimationID):
* Source/WebKit/WebKitSwift/TextAnimation/TextAnimationManager.swift:
(beginEffect(for:style:)):
(TextAnimationManager.performReplacementAndGeneratePreview(for:effect:animation:)):
* Source/WebKit/WebKitSwift/TextAnimation/WKSTextAnimationSourceDelegate.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::addSourceTextAnimationForActiveWritingToolsSession):
(WebKit::WebChromeClient::addDestinationTextAnimationForActiveWritingToolsSession):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
(WebKit::TextAnimationController::addInitialTextAnimationForActiveWritingToolsSession):
(WebKit::TextAnimationController::addSourceTextAnimationForActiveWritingToolsSession):
(WebKit::TextAnimationController::addDestinationTextAnimationForActiveWritingToolsSession):
(WebKit::TextAnimationController::updateUnderlyingTextVisibilityForTextAnimationID):
(WebKit::TextAnimationController::createTextIndicatorForRange):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::addSourceTextAnimationForActiveWritingToolsSession):
(WebKit::WebPage::addDestinationTextAnimationForActiveWritingToolsSession):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, SuggestedTextIsSelectedAfterSmartReply)):

Canonical link: <a href="https://commits.webkit.org/283248@main">https://commits.webkit.org/283248@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/98f7458b0d5ababe8b7d101d816c200e3cede7eb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65693 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45066 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18312 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69719 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16302 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67811 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52865 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16584 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/52737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11310 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68760 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41605 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56856 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/33365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38281 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14236 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15178 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/60121 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14578 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71425 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9648 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14010 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/60052 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9680 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56922 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60329 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14477 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7961 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1613 "Passed tests") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/40874 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41950 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/43133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41694 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/dom/Document.h:
```diff
@@ -1898,6 +1899,12 @@ class Document
+    void setPlaceholderTextIndicatorData(TextIndicatorData data) { m_placeholderTextIndicatorData = data; }
+    std::optional<TextIndicatorData> placeholderTextIndicatorData() { return m_placeholderTextIndicatorData; }
+#endif
+
```
```suggestion
```

pxlcoder commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -3679,6 +3679,23 @@ RefPtr<TextPlaceholderElement> Editor::insertTextPlaceholder(const IntSize& size
 
     document->selection().setSelection(VisibleSelection { positionInParentBeforeNode(placeholder.ptr()) }, FrameSelection::defaultSetSelectionOptions(UserTriggered::Yes));
 
+    // For Writing Tools, we need the snapshot of the last inserted placeholder.
```
Since this is just for Writing Tools, I think we should move this logic into the controller.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -4752,9 +4740,9 @@ - (NSUUID *)_enableSourceTextAnimationAfterElementWithID:(NSString *)elementID
 
 #if PLATFORM(IOS_FAMILY)
-    [_contentView addTextAnimationForAnimationID:nsUUID.get() withStyleType:WKTextAnimationTypeInitial];
+    [_contentView addTextAnimationForAnimationID:nsUUID.get() withData: { WebCore::TextAnimationType::Initial, WebCore::TextAnimationRunMode::RunAnimation }];
```
```suggestion
    [_contentView addTextAnimationForAnimationID:nsUUID.get() withData:{ WebCore::TextAnimationType::Initial, WebCore::TextAnimationRunMode::RunAnimation }];
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -4775,9 +4763,9 @@ - (NSUUID *)_enableFinalTextAnimationForElementWithID:(NSString *)elementID
 
 #if PLATFORM(IOS_FAMILY)
-    [_contentView addTextAnimationForAnimationID:nsUUID.get() withStyleType:WKTextAnimationTypeFinal];
+    [_contentView addTextAnimationForAnimationID:nsUUID.get() withData: { WebCore::TextAnimationType::Final, WebCore::TextAnimationRunMode::RunAnimation }];
```
```suggestion
    [_contentView addTextAnimationForAnimationID:nsUUID.get() withData:{ WebCore::TextAnimationType::Final, WebCore::TextAnimationRunMode::RunAnimation }];
```

pxlcoder commented on Source/WebKit/UIProcess/WebPageProxyInternals.h:
```diff
@@ -345,6 +345,8 @@ struct WebPageProxy::Internals final : WebPopupMenuProxy::Client
     HashMap<WTF::UUID, WebCore::TextIndicatorData> textIndicatorDataForAnimationID;
     HashMap<WTF::UUID, CompletionHandler<void(WebCore::TextAnimationRunMode)>> completionHandlerForAnimationID;
+    HashMap<WTF::UUID, CompletionHandler<void(std::optional<WebCore::TextIndicatorData>)>> completionHandlerForDestinationTextIndicatorForSourceID;
+    HashMap<WTF::UUID, WTF::UUID> sourceIDtoDestiationID;
```
```suggestion
    HashMap<WTF::UUID, WTF::UUID> sourceIDtoDestinationID;
```

pxlcoder replied:
May also want a name that has "animation" in it, as this is just a member of `WebPageProxyInternals`.

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -13303,7 +13305,11 @@ - (void)targetedPreviewForID:(NSUUID *)uuid completionHandler:(void (^)(UITarget
 
 - (void)updateUnderlyingTextVisibilityForTextAnimationID:(NSUUID *)uuid visible:(BOOL)visible completionHandler:(void (^)(void))completionHandler
 {
+    NSUUID* destinationUUID = _sourceAnimationIDtoDestinationAnimationID.get()[uuid];
```
```suggestion
    NSUUID *destinationUUID = _sourceAnimationIDtoDestinationAnimationID.get()[uuid];
```

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -13315,9 +13321,51 @@ - (UIView *)containingViewForTextAnimationType
 
 - (void)callCompletionHandlerForAnimationID:(NSUUID *)uuid
+{
+    NSUUID* destinationUUID = _sourceAnimationIDtoDestinationAnimationID.get()[uuid];
```
```suggestion
    NSUUID *destinationUUID = _sourceAnimationIDtoDestinationAnimationID.get()[uuid];
```

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -13315,9 +13321,51 @@ - (UIView *)containingViewForTextAnimationType
 {
     auto animationUUID = WTF::UUID::fromNSUUID(uuid);
+
+    // store this completion handler so that it can be called after the execution of the next
```
```suggestion
    // Store this completion handler so that it can be called after the execution of the next
```

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
```diff
@@ -310,18 +332,22 @@
 
 std::optional<WebCore::TextIndicatorData> TextAnimationController::createTextIndicatorForRange(const WebCore::SimpleRange& range)
 {
+    auto rangeForIndicator = range;
```
Why the copy?

megangardner replied:
Oh, this is leftover from when I was trying to grab the snapshot of the placeholder here.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
```diff
@@ -196,26 +199,23 @@
+
+    if (finished)
+        runMode = WebCore::TextAnimationRunMode::OnlyReplaceText;
+
```
```suggestion
```

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
```diff
@@ -32,13 +32,16 @@
 #include <WebCore/TextIndicator.h>
 #include <WebCore/TextIterator.h>
+#include <WebCore/TextPlaceholderElement.h>
+#include <WebCore/TypedElementDescendantIteratorInlines.h>
```
I don't think any of these includes are necessary?

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
```diff
@@ -51,8 +51,14 @@
+
 #import <pal/cocoa/WritingToolsUISoftLink.h>
 
+
```
```suggestion
```

rr-codes commented on Source/WebCore/page/TextAnimationTypes.h:
```diff
@@ -45,6 +45,8 @@ struct TextAnimationData {
     TextAnimationType style;
     TextAnimationRunMode runMode;
     WTF::UUID unanimatedRangeUUID { WTF::UUID::emptyValue };
+    WTF::UUID sourceAnimationUUID { WTF::UUID::emptyValue };
```
can these be ObjectIdentifiers?

megangardner replied:
Everything runs on UUIDs right now, I don't want to change that at this point, but we're free to do a full changeover in the future!

rr-codes commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1268,6 +1268,26 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+
+    // Storing and sending information for the different shaped SPI on iOS.
+    if (styleData.runMode == WebCore::TextAnimationRunMode::RunAnimation) {
+        if (styleData.style == WebCore::TextAnimationType::Source)
```
can this be made into a switch-case over `styleData.style`

megangardner replied:
I'm not sure what that would get us. We only want to do these things for the animations types, we don't care if it's the initial animation.

rr-codes commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
```diff
@@ -565,6 +565,7 @@ struct ImageAnalysisContextMenuActionData {
     RetainPtr<UIView<UITextCursorView>> _editDropTextCursorView;
     RetainPtr<UITextCursorDropPositionAnimator> _editDropCaretAnimator;
 #endif
+    RetainPtr<NSMutableDictionary> _sourceAnimationIDtoDestinationAnimationID;
```
please add the generic type parameters to `NSMutableDictionary`

rr-codes commented on Source/WebCore/page/ChromeClient.h:
```diff
@@ -683,9 +683,11 @@ class ChromeClient {
     virtual void addInitialTextAnimationForActiveWritingToolsSession() { }
 
-    virtual void addSourceTextAnimationForActiveWritingToolsSession(const CharacterRange&, const String&, CompletionHandler<void(TextAnimationRunMode)>&&) { }
+    virtual void addSourceTextAnimationForActiveWritingToolsSession(const WTF::UUID&, const WTF::UUID&, bool, const CharacterRange&, const String&, CompletionHandler<void(TextAnimationRunMode)>&&) { }
```
these parameters should have names

rr-codes commented on Source/WebCore/page/ChromeClient.h:
```diff
@@ -683,9 +683,11 @@ class ChromeClient {
+    virtual void addSourceTextAnimationForActiveWritingToolsSession(const WTF::UUID&, const WTF::UUID&, bool, const CharacterRange&, const String&, CompletionHandler<void(TextAnimationRunMode)>&&) { }
 
-    virtual void addDestinationTextAnimationForActiveWritingToolsSession(const std::optional<CharacterRange>&, const String&) { }
+    virtual void addDestinationTextAnimationForActiveWritingToolsSession(const WTF::UUID&, const WTF::UUID&, const std::optional<CharacterRange>&, const String&) { }
```
ditto

rr-codes commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -568,11 +567,11 @@
     auto rangeAfterReplace = characterRange(sessionRange, *selectionRange);
 
-    compositionSessionDidFinishReplacement(rangeAfterReplace, attributedText.string);
+    compositionSessionDidFinishReplacement(sourceAnimationUUID, destinationAnimationUUID, rangeAfterReplace, attributedText.string);
```
when is there ever more than one destination / source animation ongoing at the same time?

megangardner replied:
I'm not sure I understand the question. We need these IDs so that the destination animation can run properly on iOS.

rr-codes commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -584,15 +583,21 @@
+    if (session.compositionType == WritingTools::SessionCompositionType::SmartReply)
+        finished = false;
+
+    m_page->chrome().client().addSourceTextAnimationForActiveWritingToolsSession(sourceAnimationUUID, destinationAnimationUUID, finished, range, attributedText.string, WTFMove(addDestinationTextAnimation));
```
can't the uuids be made within `addSourceTextAnimationForActiveWritingToolsSession`, and then just call `addDestinationTextAnimation` with those as the parameters?

megangardner replied:
No, we need them to be passed in to each function call, and to be created outside of completion handlers. I tried to do what you describe, but it didn't work.

rr-codes commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1268,6 +1268,26 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+
+        if (styleData.style == WebCore::TextAnimationType::Final) {
+            if (auto sourceAnimationID = internals().sourceAnimationIDtoDestinationAnimationID.take(uuid)) {
+                if (auto completionHandler = internals().completionHandlerForDestinationTextIndicatorForSourceID.take(sourceAnimationID))
```
what clears completionHandlerForDestinationTextIndicatorForSourceID and sourceAnimationIDtoDestinationAnimationID ?

megangardner replied:
Nothing, these are leaked at this time.

pxlcoder replied:
But we shouldn't be adding leaks??

rr-codes commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1280,6 +1300,13 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
 }
 
+#if PLATFORM(IOS_FAMILY)
+void WebPageProxy::storeDestinationCompletionHandlerForAnimationID(const WTF::UUID& uuid, CompletionHandler<void(std::optional<WebCore::TextIndicatorData>)>&& completionHandler)
```
`uuid` -> `sourceAnimationID`

megangardner replied:
I think you mean destination, but sure.

rr-codes commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -13315,9 +13321,51 @@ - (UIView *)containingViewForTextAnimationType
+    // call to replace the text and eventually use this completion handler to pass the
+    // text indicator to UIKit.
+    _page->storeDestinationCompletionHandlerForAnimationID(*animationUUID, [protectedSelf = retainPtr(self),  completionHandler = makeBlockPtr(completionHandler)] (std::optional<WebCore::TextIndicatorData> textIndicatorData) {
+
```
extra newline

rr-codes commented on Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
```diff
@@ -177,7 +177,7 @@ - (void)restoreTextAnimationType
         RetainPtr effectData = [_chunkToEffect objectForKey:chunkID];
         if ([effectData type] == WebCore::TextAnimationType::Initial)
-            [self addTextAnimationForAnimationID:chunkID withData: { WebCore::TextAnimationType::Initial, WebCore::TextAnimationRunMode::RunAnimation, WTF::UUID(WTF::UUID::emptyValue) }];
+            [self addTextAnimationForAnimationID:chunkID withData: { WebCore::TextAnimationType::Initial, WebCore::TextAnimationRunMode::RunAnimation }];
```
```suggestion
            [self addTextAnimationForAnimationID:chunkID withData:{ WebCore::TextAnimationType::Initial, WebCore::TextAnimationRunMode::RunAnimation }];
```

rr-codes commented on Source/WebKit/WebKitSwift/TextAnimation/TextAnimationManager.swift:
```diff
@@ -45,10 +45,11 @@ import WebKitSwift
             let newEffect = self.effectView.addEffect(UITextEffectView.PonderingEffect(chunk: TextEffectChunk(uuid: uuid), view: self.effectView) as UITextEffectView.TextEffect)
             self.chunkToEffect[uuid] = newEffect
-        } else {
+        } else if style == .source {
```
this should be a switch case

rr-codes commented on Source/WebKit/WebKitSwift/TextAnimation/TextAnimationManager.swift:
```diff
@@ -45,10 +45,11 @@ import WebKitSwift
             let newEffect = self.effectView.addEffect(UITextEffectView.ReplacementTextEffect(chunk: TextEffectChunk(uuid: uuid), view: self.effectView, delegate:self) as UITextEffectView.TextEffect)
             self.chunkToEffect[uuid] = newEffect
         }
+        // Drop .final on the floor since we don't manually start the 2nd part of the animation on iOS.
```
```suggestion
        // Discard `.final` since we don't manually start the 2nd part of the animation on iOS.
```

rr-codes commented on Source/WebKit/WebKitSwift/TextAnimation/TextAnimationManager.swift:
```diff
@@ -107,9 +108,10 @@ extension TextAnimationManager: UITextEffectView.ReplacementTextEffect.Delegate
+        let preview = await withCheckedContinuation { continuation in
+            delegate.callCompletionHandler(forAnimationID: uuidChunk.uuid, completionHandler: { preview in
+            continuation.resume(returning: preview)
+            })
```
```suggestion
        let preview = await withCheckedContinuation { continuation in
            delegate.callCompletionHandler(forAnimationID: uuidChunk.uuid) { preview in
                continuation.resume(returning: preview)
            }
        }
```

rr-codes commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -1892,16 +1892,22 @@ void WebChromeClient::addInitialTextAnimationForActiveWritingToolsSession()
 }
 
-void WebChromeClient::addDestinationTextAnimationForActiveWritingToolsSession(const std::optional<CharacterRange>& range, const String& string)
+void WebChromeClient::addDestinationTextAnimationForActiveWritingToolsSession(const WTF::UUID& sourceAnimationUUID, const WTF::UUID& destinationAnimationUUID, const std::optional<CharacterRange>& range, const String& string)
```
```suggestion
void WebChromeClient::addDestinationTextAnimationForActiveWritingToolsSession(const WTF::UUID& sourceAnimationID, const WTF::UUID& destinationAnimationID, const std::optional<CharacterRange>& range, const String& string)
```

megangardner replied:
This is like everywhere. If we want to change it, lets do it at the same time that we change to ObjectIDs.

rr-codes commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
```diff
@@ -524,9 +524,11 @@ class WebChromeClient final : public WebCore::ChromeClient {
     void addInitialTextAnimationForActiveWritingToolsSession() final;
 
-    void addSourceTextAnimationForActiveWritingToolsSession(const WebCore::CharacterRange&, const String&, CompletionHandler<void(WebCore::TextAnimationRunMode)>&&) final;
+    void addSourceTextAnimationForActiveWritingToolsSession(const WTF::UUID& sourceAnimationUUID, const WTF::UUID& destinationAnimationUUID, bool finished, const WebCore::CharacterRange&, const String&, CompletionHandler<void(WebCore::TextAnimationRunMode)>&&) final;
```
ditto about changing the `UUID` in the parameter names to `ID`

megangardner replied:
This is like everywhere. If we want to change it, lets do it at the same time that we change to ObjectIDs.

pxlcoder commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -3679,6 +3681,14 @@ RefPtr<TextPlaceholderElement> Editor::insertTextPlaceholder(const IntSize& size
+    // For Writing Tools, we need the snapshot of the last inserted placeholder.
+    auto placeholderRange = makeRangeSelectingNode(placeholder.get());
+
+    if (placeholderRange)
```
```suggestion
    if (auto placeholderRange = makeRangeSelectingNode(placeholder.get()))
```

pxlcoder commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -3679,6 +3681,14 @@ RefPtr<TextPlaceholderElement> Editor::insertTextPlaceholder(const IntSize& size
+    auto placeholderRange = makeRangeSelectingNode(placeholder.get());
+
+    if (placeholderRange)
+        protectedDocument()->page()->chrome().client().saveSnapshotOfTextPlaceholder(*placeholderRange);
```
Rename to `saveSnapshotOfTextPlaceholderForAnimation`?

pxlcoder commented on Source/WebCore/page/ChromeClient.h:
```diff
@@ -683,9 +683,11 @@ class ChromeClient {
     virtual void addInitialTextAnimationForActiveWritingToolsSession() { }
 
-    virtual void addSourceTextAnimationForActiveWritingToolsSession(const CharacterRange&, const String&, CompletionHandler<void(TextAnimationRunMode)>&&) { }
+    virtual void addSourceTextAnimationForActiveWritingToolsSession(const WTF::UUID& /*sourceAnimationUUID*/, const WTF::UUID& /*destinationAnimationUUID*/, bool, const CharacterRange&, const String&, CompletionHandler<void(TextAnimationRunMode)>&&) { }
```
You can remove the `/* */` and simply name the variable.

megangardner replied:
Other variables with names in this file are done with comments, so sticking with that.

rr-codes replied:
(it should be `/* destinationAnimationUUID */` instead of `/*destinationAnimationUUID*/`

pxlcoder commented on Source/WebCore/page/ChromeClient.h:
```diff
@@ -683,9 +683,11 @@ class ChromeClient {
+    virtual void addSourceTextAnimationForActiveWritingToolsSession(const WTF::UUID& /*sourceAnimationUUID*/, const WTF::UUID& /*destinationAnimationUUID*/, bool, const CharacterRange&, const String&, CompletionHandler<void(TextAnimationRunMode)>&&) { }
 
-    virtual void addDestinationTextAnimationForActiveWritingToolsSession(const std::optional<CharacterRange>&, const String&) { }
+    virtual void addDestinationTextAnimationForActiveWritingToolsSession(const WTF::UUID& /*sourceAnimationUUID*/, const WTF::UUID& /*destinationAnimationUUID*/, const std::optional<CharacterRange>&, const String&) { }
```
Ditto.

megangardner replied:
ditto as well

pxlcoder commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -584,15 +583,21 @@
+    auto destinationAnimationUUID = WTF::UUID::createVersion4();
+
+    auto addDestinationTextAnimation = [weakThis = WeakPtr { *this }, attributedText, range, context, finished, sourceAnimationUUID, destinationAnimationUUID](TextAnimationRunMode runMode) mutable {
+        weakThis->compositionSessionDidReceiveTextWithReplacementRangeAsync(sourceAnimationUUID, destinationAnimationUUID, attributedText, range, context, finished, runMode);
```
Need to null check `weakThis`?

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -13396,15 +13443,30 @@ - (void)writingToolsSession:(WTSession *)session didReceiveAction:(WTAction)acti
+- (void)addTextAnimationForAnimationID:(NSUUID *)uuid withData:(const WebCore::TextAnimationData&)data
+{
+    if (data.style == WebCore::TextAnimationType::Final)
+        [_sourceAnimationIDtoDestinationAnimationID setObject:uuid forKey:data.sourceAnimationUUID];
```
This should be moved after the early return due to the preference being disabled.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
```diff
@@ -225,28 +222,37 @@
+        if (m_activeAnimation) {
+            if (m_finalReplaceHandler) {
+                (*m_finalReplaceHandler)(WebCore::TextAnimationRunMode::OnlyReplaceText);
+                m_finalReplaceHandler = std::nullopt;
```
Should be able to use `std::exchange` here.

pxlcoder replied:
Seems we can actually remove the writing of nullopt entirely.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
```diff
@@ -333,6 +368,12 @@
 
+    // For smart replies, we need to use a blank snapshot
+    if (range.collapsed()) {
+        if (m_placeholderTextIndicatorData)
```
This can be combined into a single if.

pxlcoder commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -1892,16 +1892,22 @@ void WebChromeClient::addInitialTextAnimationForActiveWritingToolsSession()
+    protectedPage()->saveSnapshotOfTextPlaceholder(placeholderRange);
+}
+
+
```
```suggestion
```

pxlcoder commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -584,15 +583,21 @@
+        finished = false;
+
+    m_page->chrome().client().addSourceTextAnimationForActiveWritingToolsSession(sourceAnimationUUID, destinationAnimationUUID, finished, range, attributedText.string, WTFMove(addDestinationTextAnimation));
+
```
```suggestion
```

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
```diff
@@ -333,6 +366,10 @@
         return std::nullopt;
     }
 
+    // For smart replies, we need to use a blank snapshot
```
```suggestion
    // For smart replies, we need to use a blank snapshot.
```

rr-codes commented on Source/WebCore/page/ChromeClient.h:
```diff
@@ -683,9 +683,11 @@ class ChromeClient {
     virtual void addInitialTextAnimationForActiveWritingToolsSession() { }
 
-    virtual void addSourceTextAnimationForActiveWritingToolsSession(const CharacterRange&, const String&, CompletionHandler<void(TextAnimationRunMode)>&&) { }
+    virtual void addSourceTextAnimationForActiveWritingToolsSession(const WTF::UUID& /*sourceAnimationUUID*/, const WTF::UUID& /*destinationAnimationUUID*/, bool, const CharacterRange&, const String&, CompletionHandler<void(TextAnimationRunMode)>&&) { }
```
I think the bool parameter should also be labelled

rr-codes commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -550,15 +549,21 @@
 
-    auto addDestinationTextAnimation = [weakThis = WeakPtr { *this }, attributedText, range, context, finished](TextAnimationRunMode runMode) mutable {
-        weakThis->compositionSessionDidReceiveTextWithReplacementRangeAsync(attributedText, range, context, finished, runMode);
+    // Must generate these UUID now to pass into the source animation for iOS to work.
```
`UUID` -> `UUIDs`

rr-codes commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -550,15 +549,21 @@
-#endif
+    // We only get a single replace call for smart replies with finished = true. We use this flag to not run the final replace for a compsition
+    // session, so for smart replies, we need to make sure to not send with this flag, so that we can be sure to do the animation for smart replies.
+    if (session.compositionType == WritingTools::SessionCompositionType::SmartReply)
+        finished = false;
```
```suggestion
    auto performFinalReplace = finished && session.compositionType != WritingTools::SessionCompositionType::SmartReply; 
```

rr-codes commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -550,15 +549,21 @@
-#else
-    addDestinationTextAnimation(WebCore::TextAnimationRunMode::RunAnimation);
-#endif
+    // We only get a single replace call for smart replies with finished = true. We use this flag to not run the final replace for a compsition
```
compsition -> composition

rr-codes commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -13315,9 +13321,50 @@ - (UIView *)containingViewForTextAnimationType
+    // Store this completion handler so that it can be called after the execution of the next
+    // call to replace the text and eventually use this completion handler to pass the
+    // text indicator to UIKit.
+    _page->storeDestinationCompletionHandlerForAnimationID(*animationUUID, [protectedSelf = retainPtr(self),  completionHandler = makeBlockPtr(completionHandler)] (std::optional<WebCore::TextIndicatorData> textIndicatorData) {
```
there's two space before completionHandler instead of one

megangardner committed:
Writing tools animations on iOS don't have the second part of the animation.
https://bugs.webkit.org/show_bug.cgi?id=279161
rdar://132367375

Reviewed by Aditya Keerthi.

iOS animations have a completely different shape than mac
animations. Also, we weren't correctly determining if something
was the final text replacement and bailing out of that animations.
We also needed to ensure that smart replies still worked correctly.

* Source/WebCore/dom/Document.h:
(WebCore::Document::setPlaceholderTextIndicatorData):
(WebCore::Document::placeholderTextIndicatorData):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::insertTextPlaceholder):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::addSourceTextAnimationForActiveWritingToolsSession):
(WebCore::ChromeClient::addDestinationTextAnimationForActiveWritingToolsSession):
* Source/WebCore/page/TextAnimationTypes.h:
* Source/WebCore/page/writing-tools/WritingToolsController.h:
* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::compositionSessionDidFinishReplacement):
(WebCore::WritingToolsController::compositionSessionDidReceiveTextWithReplacementRangeAsync):
(WebCore::WritingToolsController::compositionSessionDidReceiveTextWithReplacementRange):
* Source/WebKit/Shared/TextAnimationTypes.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _addTextAnimationForAnimationID:withData:]):
(-[WKWebView _enableSourceTextAnimationAfterElementWithID:]):
(-[WKWebView _enableFinalTextAnimationForElementWithID:]):
(toWKTextAnimationType): Deleted.
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::addTextAnimationForAnimationIDWithCompletionHandler):
(WebKit::WebPageProxy::storeDestinationCompletionHandlerForAnimationID):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setUpInteraction]):
(-[WKContentView updateUnderlyingTextVisibilityForTextAnimationID:visible:completionHandler:]):
(-[WKContentView callCompletionHandlerForAnimationID:]):
(-[WKContentView callCompletionHandlerForAnimationID:completionHandler:]):
(toWKTextAnimationType):
(-[WKContentView addTextAnimationForAnimationID:withData:]):
(-[WKContentView addTextAnimationForAnimationID:withStyleType:]): Deleted.
* Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
(-[WKTextAnimationManager restoreTextAnimationType]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::addTextAnimationForAnimationID):
* Source/WebKit/WebKitSwift/TextAnimation/TextAnimationManager.swift:
(beginEffect(for:style:)):
(TextAnimationManager.performReplacementAndGeneratePreview(for:effect:animation:)):
* Source/WebKit/WebKitSwift/TextAnimation/WKSTextAnimationSourceDelegate.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::addSourceTextAnimationForActiveWritingToolsSession):
(WebKit::WebChromeClient::addDestinationTextAnimationForActiveWritingToolsSession):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
(WebKit::TextAnimationController::addInitialTextAnimationForActiveWritingToolsSession):
(WebKit::TextAnimationController::addSourceTextAnimationForActiveWritingToolsSession):
(WebKit::TextAnimationController::addDestinationTextAnimationForActiveWritingToolsSession):
(WebKit::TextAnimationController::updateUnderlyingTextVisibilityForTextAnimationID):
(WebKit::TextAnimationController::createTextIndicatorForRange):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::addSourceTextAnimationForActiveWritingToolsSession):
(WebKit::WebPage::addDestinationTextAnimationForActiveWritingToolsSession):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, SuggestedTextIsSelectedAfterSmartReply)):

Canonical link: https://commits.webkit.org/283248@main
